File: cpp/64.Minimum-Path-Sum.DP.cpp
// 98% faster
// O(m*n) time, O(m*n) space

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int table[grid.size()][grid[0].size()];
        table[grid.size() - 1][grid[0].size() - 1] = grid[grid.size() - 1][grid[0].size() - 1];
        for(int i{int(grid.size()) - 2}; i>=0; i--){
            table[i][grid[0].size() - 1] = grid[i][grid[0].size() - 1] + table[i + 1][grid[0].size() - 1];
        }
        for(int i{int(grid[0].size()) - 2};i>=0;i--){
            table[grid.size() - 1][i] = grid[grid.size() - 1][i] + table[grid.size() - 1][i+1];
        }
        for(int i{int(grid.size()) - 2}; i>=0; i--){
            for(int j{int(grid[0].size()) - 2}; j>=0; j--){
                table[i][j] = grid[i][j] + min(table[i][j+1], table[i+1][j]);
            }
        }
        return table[0][0];
    }
};

File: Java/64. Minimum Path Sum.DP.java
class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (i == 0 && j == 0) {
                    continue;
                }
                if (i == 0) {
                    grid[i][j] += grid[i][j - 1];
                    continue;
                }
                if (j == 0) {
                    grid[i][j] += grid[i - 1][j];
                    continue;
                }
                grid[i][j] += Math.min(grid[i - 1][j], grid[i][j - 1]);
            }
        }
        return grid[m - 1][n - 1];
    }
}
// time complexity : faster 84.99% O(m * n)
// space complexity : faster 67.33% O(m * n)
